components/MakeFunction.coffee

```diff
@@ -1,61 +1,52 @@
 noflo = require 'noflo'
 
-class MakeFunction extends noflo.Component
-  description: 'Evaluates a function each time data hits the "in" port
+exports.getComponent = ->
+  c = new noflo.Component
+  c.description = 'Evaluates a function each time data hits the "in" port
   and sends the return value to "out". Within the function "x" will
   be the variable from the in port. For example, to make a ^2 function
   input "return x*x;" to the function port.'
-  icon: 'code'
-
-  constructor: ->
-    @f = null
-
-    # We have two input ports. One for the callback to call, and one
-    # for IPs to call it with
-    @inPorts = new noflo.InPorts
-      in:
-        datatype: 'all'
-        description: 'Packet to be processed'
-      function:
-        datatype: 'string'
-        description: 'Function to evaluate'
-    # The optional error port is used in case of wrong setups
-    @outPorts = new noflo.OutPorts
-      out:
-        datatype: 'all'
-      function:
-        datatype: 'function'
-      error:
-        datatype: 'object'
-
-    # Set callback
-    @inPorts.function.on 'data', (data) =>
-      if typeof data is "function"
-        @f = data
-      else
+  c.icon = 'code'
+
+  c.inPorts.add 'in',
+    datatype: 'all'
+    description: 'Packet to be processed'
+  c.inPorts.add 'function',
+    datatype: 'string'
+    description: 'Function to evaluate'
+    control: true
+  c.outPorts.add 'out',
+    datatype: 'all'
+  c.outPorts.add 'function',
+    datatype: 'function'
+  c.outPorts.add 'error',
+    datatype: 'object'
+
+  c.process (input, output) ->
+    if input.has 'function'
+      func = input.getData 'function'
+      unless typeof func is 'function'
         try
-          @f = Function("x", data)
-        catch error
-          @error 'Error creating function: ' + data
-      if @f and @outPorts.function.isAttached()
-        @outPorts.function.send @f
-
-
-    # Evaluate the function when receiving data
-    @inPorts.in.on 'data', (data) =>
-      unless @f
-        @error 'No function defined'
+          func = Function 'x', func
+        catch e
+          output.sendDone e
+          return
+      unless input.has 'in'
```

Comment:
What if `in` arrives a bit later than `function`? Then `function` would be sent as is and `in` would be queued until the next `function` arrives. Should probably be synchronized explicitly in the graph, but some might expect implicit synchronization here.